Physical camera fixes

Thomasch-unity3d opened the pull request:
### Purpose of this PR
Adds Physical Camera properties UI to the HDCamera inspector and fixes procedural sky and HDRI sky when using physical camera mode with non-zero lens shift. This PR relies on methods not available yet in released version of Unity.
---
### Release Notes
- Enabled Physical Camera mode in HDRP.

---
### Testing status
Tested manually, waiting for QA pass by joelf.
---
### Overall Product Risks
**Technical Risk**: Low

**Halo Effect**: Medium

---
### Comments to reviewers
Do not merge, waiting for changes to land in 2018.3 first.

sebastienlagarde commented:
Adding Fredéric and Remi if they have feedback.

fredpointzero commented on com.unity.render-pipelines.core/CoreRP/Utilities/GeometryUtils.cs:
```diff
@@ -255,7 +255,7 @@ public static Matrix4x4 CalculateProjectionMatrix(Camera camera)
             }
             else
-                return Matrix4x4.Perspective(camera.fieldOfView, camera.aspect, camera.nearClipPlane, camera.farClipPlane);
+                return Matrix4x4.Perspective(camera.GetGateFittedFOV(), camera.aspect, camera.nearClipPlane, camera.farClipPlane);
```
Do we always use physical properties here? Or should it depends on the camera.usePhysicalProperties property?

Thomasch-unity3d replied:
GetGateFittedFOV returns the fov when the camera is not physical.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -162,34 +162,17 @@ static void Drawer_FieldVolumeAnchorOverride(HDCameraUI s, SerializedHDCamera p,
 
         static void Drawer_FieldCullingMask(HDCameraUI s, SerializedHDCamera p, Editor owner)
         {
-            EditorGUILayout.PropertyField(p.cullingMask, _.GetContent("Culling Mask"));
```
Keep the UI code of the HDRP for all drawers, because:
 - The UI is not shared exactly as is in all SRPs
 - It is a call to a black box implementation. (If we have specific HDRP UI/UX requests, we can't change it).

So here, it should be (if I understand properly `p.baseCameraSettings`):
`EditorGUILayout.PropertyField(p.baseCameraSettings.cullingMask, _.GetContent("Culling Mask"));`

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -162,34 +162,17 @@ static void Drawer_FieldVolumeAnchorOverride(HDCameraUI s, SerializedHDCamera p,
-                else
-                    EditorGUILayout.Slider(p.fieldOfView, 1f, 179f, _.GetContent("Field of View|The width of the Camera’s view angle, measured in degrees along the local Y axis."));
-            }
+            p.baseCameraSettings.DrawProjection();
```
Same request

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -162,34 +162,17 @@ static void Drawer_FieldVolumeAnchorOverride(HDCameraUI s, SerializedHDCamera p,
-                "Clipping Planes",
-                new[] { p.nearClippingPlane, p.farClippingPlane },
-                new[] { _.GetContent("Near|The closest point relative to the camera that drawing will occur."), _.GetContent("Far|The furthest point relative to the camera that drawing will occur.\n") });
+            p.baseCameraSettings.DrawClippingPlanes();
```
Same request

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -211,12 +194,12 @@ static void Drawer_FieldIso(HDCameraUI s, SerializedHDCamera p, Editor owner)
         static void Drawer_FieldNormalizedViewPort(HDCameraUI s, SerializedHDCamera p, Editor owner)
         {
-            EditorGUILayout.PropertyField(p.normalizedViewPortRect, _.GetContent("Viewport Rect|Four values that indicate where on the screen this camera view will be drawn. Measured in Viewport Coordinates (values 0–1)."));
+            p.baseCameraSettings.DrawNormalizedViewPort();
```
Same request

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -211,12 +194,12 @@ static void Drawer_FieldIso(HDCameraUI s, SerializedHDCamera p, Editor owner)
         static void Drawer_FieldDepth(HDCameraUI s, SerializedHDCamera p, Editor owner)
         {
-            EditorGUILayout.PropertyField(p.depth, _.GetContent("Depth"));
+            p.baseCameraSettings.DrawDepth();
```
Same request

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraEditor.Handlers.cs:
```diff
@@ -21,7 +21,7 @@ void OnSceneGUI()
             SceneViewOverlay_Window(_.GetContent("Camera Preview"), OnOverlayGUI, -100, target);
 
-            CameraEditorUtils.HandleFrustrum(c);
+            UnityEditor.CameraEditorUtils.HandleFrustum(c, c.GetInstanceID());
```
Thanks for the update with the `UnityEditor.CameraEditorUtils` class

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -236,34 +219,19 @@ static void Drawer_FieldClearDepth(HDCameraUI s, SerializedHDCamera p, Editor ow
-                    EditorGUILayout.HelpBox("Manual MSAA target set with deferred rendering. This will lead to undefined behavior.", MessageType.Warning, true);
-                }
-            }
+            p.baseCameraSettings.DrawTargetTexture(!p.frameSettings.enableForwardRenderingOnly.boolValue);
```
Same request

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -236,34 +219,19 @@ static void Drawer_FieldClearDepth(HDCameraUI s, SerializedHDCamera p, Editor ow
         static void Drawer_FieldOcclusionCulling(HDCameraUI s, SerializedHDCamera p, Editor owner)
         {
-            EditorGUILayout.PropertyField(p.occlusionCulling, _.GetContent("Occlusion Culling"));
+            p.baseCameraSettings.DrawOcclusionCulling();
```
Same request.

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -236,34 +219,19 @@ static void Drawer_FieldClearDepth(HDCameraUI s, SerializedHDCamera p, Editor ow
-                var warnings = GetCameraBufferWarnings(camera);
-                if (warnings.Length > 0)
-                    EditorGUILayout.HelpBox(string.Join("\n\n", warnings), MessageType.Warning, true);
+                CameraEditor.Settings.DrawCameraWarnings(camera);
```
Same request.

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -280,35 +248,21 @@ static void Drawer_FieldVR(HDCameraUI s, SerializedHDCamera p, Editor owner)
             }
-            EditorGUILayout.PropertyField(p.stereoSeparation, _.GetContent("Stereo Separation"));
-            EditorGUILayout.PropertyField(p.stereoConvergence, _.GetContent("Stereo Convergence"));
+            EditorGUILayout.PropertyField(p.baseCameraSettings.stereoSeparation, _.GetContent("Stereo Separation"));
```
Same request.

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -280,35 +248,21 @@ static void Drawer_FieldVR(HDCameraUI s, SerializedHDCamera p, Editor owner)
-                if (prevDisplay != p.targetDisplay.intValue)
-                    UnityEditorInternal.InternalEditorUtility.RepaintAllViews();
-            }
+            p.baseCameraSettings.DrawMultiDisplay();
```
Same request.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/HDCameraUI.cs:
```diff
@@ -280,35 +248,21 @@ static void Drawer_FieldVR(HDCameraUI s, SerializedHDCamera p, Editor owner)
         static void Drawer_FieldTargetEye(HDCameraUI s, SerializedHDCamera p, Editor owner)
         {
-            EditorGUILayout.IntPopup(p.targetEye, k_TargetEyes, k_TargetEyeValues, _.GetContent("Target Eye"));
+            p.baseCameraSettings.DrawTargetEye();
```
Same request.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/SerializedHDCamera.cs:
```diff
@@ -9,37 +9,32 @@ class SerializedHDCamera
+        private CameraEditor.Settings m_Settings;
 
+        public CameraEditor.Settings baseCameraSettings
+        {
```
I agree with @RSlysz, `baseCameraSettings`should be an auto property initialized in the constructor.

Like `public CameraEditor.Settings baseCameraSettings { get; private set; }`

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Editor/Camera/SerializedHDCamera.cs:
```diff
@@ -54,27 +49,7 @@ public SerializedHDCamera(SerializedObject serializedObject)
-#endif
-
-            targetEye = serializedObject.FindProperty("m_TargetEye");
-
```
Please, initialize `baseCameraSettings` here

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Lighting/Reflection/ReflectionSystemInternal.cs:
```diff
@@ -354,7 +354,7 @@ static void CalculateMirroredCaptureCameraProperties(PlanarReflectionProbe probe
             fov = probe.overrideFieldOfView
                 ? probe.fieldOfViewOverride
-                : Mathf.Max(viewerCamera.fieldOfView, viewerCamera.fieldOfView * viewerCamera.aspect);
+                : Mathf.Max(viewerCamera.GetGateFittedFOV(), viewerCamera.GetGateFittedFOV() * viewerCamera.aspect);
```
Should we always use the physical FOV one? Or should we choose the fov depending on `viewCamera.usePhysicalProperties`?

fredpointzero replied:
Another note: `viewerCamera.GetGateFittedFOV()`is an invalid naming for a getter for C# code.
So, if it is an actual getter (no calculation performed), you should use a C# getter.
If it does some calculation or logic, it shouldn't be prefixed with `Get`

Thomasch-unity3d replied:
GetGateFittedFOV works when physical camera mode is off. Fixing the invalid name requires a change in Unity, I'll have to take care of that later.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Lighting/Volumetrics/VolumetricLighting.cs:
```diff
@@ -537,10 +537,11 @@ public void VolumeVoxelizationPass(HDCamera hdCamera, CommandBuffer cmd, uint fr
                 Vector4 resolution  = frameParams.resolution;
-                float   vFoV        = hdCamera.camera.fieldOfView * Mathf.Deg2Rad;
+                float   vFoV        = hdCamera.camera.GetGateFittedFOV() * Mathf.Deg2Rad;
+                Vector2 lensShift   = hdCamera.camera.GetGateFittedLensShift();
```
Same comment about naming for `GetGateFittedLensShift()`

fredpointzero replied:
Please, use `var` instead of `float` or `Vector2` t odefine variables.

Thomasch-unity3d replied:
Fixed usage of var in latest commit.  Will change GetGateFittedLensShift in another PR.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Lighting/Volumetrics/VolumetricLighting.cs:
```diff
@@ -537,10 +537,11 @@ public void VolumeVoxelizationPass(HDCamera hdCamera, CommandBuffer cmd, uint fr
 
                 // Compose the matrix which allows us to compute the world space view direction.
-                Matrix4x4 transform   = HDUtils.ComputePixelCoordToWorldSpaceViewDirectionMatrix(vFoV, resolution, hdCamera.viewMatrix, false);
+                Matrix4x4 transform   = HDUtils.ComputePixelCoordToWorldSpaceViewDirectionMatrix(vFoV,lensShift, resolution, hdCamera.viewMatrix, false);
```
Missing a space before `lensShift`

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Material/GGXConvolution/RuntimeFilterIBL.cs:
```diff
@@ -130,7 +130,7 @@ public void Cleanup()
                     {
                         var faceSize = new Vector4(source.width >> mip, source.height >> mip, 1.0f / (source.width >> mip), 1.0f / (source.height >> mip));
-                        var transform = HDUtils.ComputePixelCoordToWorldSpaceViewDirectionMatrix(0.5f * Mathf.PI, faceSize, worldToViewMatrices[face], true);
+                        var transform = HDUtils.ComputePixelCoordToWorldSpaceViewDirectionMatrix(0.5f * Mathf.PI,Vector2.zero, faceSize, worldToViewMatrices[face], true);
```
Missing a space before `Vector2.zero`

Thomasch-unity3d replied:
fixed in last commit.

fredpointzero commented on com.unity.render-pipelines.high-definition/HDRP/Sky/SkyRenderingContext.cs:
```diff
@@ -95,7 +95,7 @@ public void RebuildSkyMatrices(int resolution)
                 var worldToView = lookAt * Matrix4x4.Scale(new Vector3(1.0f, 1.0f, -1.0f)); // Need to scale -1.0 on Z to match what is being done in the camera.wolrdToCameraMatrix API. ...
 
-                m_facePixelCoordToViewDirMatrices[i] = HDUtils.ComputePixelCoordToWorldSpaceViewDirectionMatrix(0.5f * Mathf.PI, m_CubemapScreenSize, worldToView, true);
+                m_facePixelCoordToViewDirMatrices[i] = HDUtils.ComputePixelCoordToWorldSpaceViewDirectionMatrix(0.5f * Mathf.PI,Vector2.zero, m_CubemapScreenSize, worldToView, true);
```
Missing a space before `Vector2.zero`

Thomasch-unity3d replied:
fixed in last commit.

sebastienlagarde commented:
A comment here. Please create a branch on public github and not on a fork as we need to launch katana on it and we can't from a fork: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=MSAA_RT_REFACTOR&automation-tools_branch=add-platform-filter&unity_branch=trunk#

sebastienlagarde commented:
@Thomasch-unity3d Sorry, I haven't follow what have happen on this PR, have the comment been adressed? @fredpointzero is it good? Can we merge latest master and move forward ?

fredpointzero commented:
@sebastienlagarde This is still not good, the issues have not been fixed.

Thomasch-unity3d commented:
Sorry guys, I didn't have time to finish that before I left last week, I'm taking care of that today.

Thomasch-unity3d commented:
I think I addressed all the PR comments and rebased onto master again, let me know if I missed anything.
The QA pass should happen today or early this week.

sebastienlagarde commented:
Thanks! Reviewing, please create a branch on this github and not in your fork as we can't launch katana test from your fork.

fredpointzero commented on com.unity.render-pipelines.high-definition/Runtime/Lighting/Reflection/ReflectionSystemInternal.cs:
```diff
@@ -523,9 +523,15 @@ static void CalculateMirroredCaptureCameraProperties(PlanarReflectionProbe probe
+#if UNITY_2019_1_OR_NEWER
+            fov = (probe.captureSettings.overrides & CaptureSettingsOverrides.FieldOfview) > 0
+                ? probe.captureSettings.fieldOfView
+                : Mathf.Max(viewerCamera.GetGateFittedFieldOfView(), viewerCamera.GetGateFittedFieldOfView() * viewerCamera.aspect);
```
`GetGateFitterFieldOfView` does not follows the C# coding standard. It should be `viewerCamera.gateFittedFieldOfView`.
But I guess this is on the Unity code side and not part of this PR.
So I won't block the PR for this, but I think it is necessary to point it out.

sebastienlagarde commented:
Side note: is this PR still working on 2018.3? is it candidate to backport? (I can do the backport)

sebastienlagarde commented:
Adding @EvgeniiG for the file volumetric + want to validate with you that macro 5OptimizeProjectionMatrix() in shader is still valid regarding this change, as maybe it will not work with the shift of the physical camera matrix.

sebastienlagarde commented:
Note: We will do a check on our various graphic featues, beause it is highly possible that physical camera parameter like "shift" simply don"t work with our approach of SSS , Volumetric, SSR, SSAO, Distortion and others... Have you done any test with these effects ON? If it is not supported it may be hard for us to support them without having a large impact on Performance, mean we will need to discuss about validity of this PR.

sebastienlagarde commented:
@RemyUnity When you are available, could you get this PR and test our various effects: Contact shadow, SSS, volumetri, refraction, SSR, distortion with a varying shift on the camera ? Goal is to see if the effect reamin stable or if there is any artifact appearing. We have done a test on postprocess V2 with legacy with THomas and both SSAO and DOF seems to remain stable under shift change. thanks

Thomasch-unity3d commented:
I did test with some effects and tested PPS2 on Legacy but likely missed some features or cases I'm not familiar with, the QA pass is still pending. Let me know if there's anything I can do to help.

sebastienlagarde commented:
FYI, I tested tile, cluster and SSS and it seems to works correctly, still need to check SSR and volumetric. Seems that SSR is working (but currently have another issue so can't really validate).

sebastienlagarde commented:
Note: I am closing this PR as I was requiring to update the code to current master + to be able to launch katana and created another PR for it: https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/2299

(please try to avoid private branch when doing feature for HDRP, make thing easier for both katana and update, thanks).
